src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs

```diff
@@ -1465,8 +1465,55 @@ internal override TypeSymbol MergeNullability(TypeSymbol other, VarianceKind var
                 hadNullabilityMismatch = false;
                 return this;
             }
-            TypeSymbol underlyingType = _underlyingType.MergeNullability(otherTuple._underlyingType, variance, out hadNullabilityMismatch);
-            return WithUnderlyingType((NamedTypeSymbol)underlyingType);
+            NamedTypeSymbol underlyingType;
+            if (MergeUnderlyingTypeNullability(_underlyingType, otherTuple._underlyingType, variance, out underlyingType, out hadNullabilityMismatch))
+            {
+                return WithUnderlyingType(underlyingType);
+            }
+            return this;
+        }
+
+        private static bool MergeUnderlyingTypeNullability(
+            NamedTypeSymbol typeA,
+            NamedTypeSymbol typeB,
+            VarianceKind variance,
+            out NamedTypeSymbol mergedType,
+            out bool hadNullabilityMismatch)
+        {
+            Debug.Assert(typeA.Equals(typeB, TypeCompareKind.IgnoreDynamicAndTupleNames | TypeCompareKind.IgnoreNullableModifiersForReferenceTypes));
+
+            hadNullabilityMismatch = false;
+            mergedType = null;
+
+            var typeDefinition = typeA.OriginalDefinition;
+            var typeParameters = typeDefinition.TypeParameters;
+            int n = typeParameters.Length;
+            var allTypeArguments = ArrayBuilder<TypeSymbolWithAnnotations>.GetInstance(n);
+
+            var typeArgumentsA = typeA.TypeArgumentsNoUseSiteDiagnostics;
+            var typeArgumentsB = typeB.TypeArgumentsNoUseSiteDiagnostics;
+            bool haveChanges = false;
+            for (int i = 0; i < n; i++)
+            {
+                TypeSymbolWithAnnotations typeArgumentA = typeArgumentsA[i];
+                TypeSymbolWithAnnotations typeArgumentB = typeArgumentsB[i];
+                TypeSymbolWithAnnotations merged = typeArgumentA.MergeNullability(typeArgumentB, variance, out bool hadMismatch);
+                hadNullabilityMismatch |= hadMismatch;
+                allTypeArguments.Add(merged);
+                if (!typeArgumentA.IsSameAs(merged))
```

Comment:
I didn't understand the use of `IsSameAs` here. That's doing reference equality, no? I expected `Equals` with some comparison options instead.

Reply:
`IsSameAs` and `hasChanges` are only used as an optimization to avoid creating unnecessary `TypeSymbol` instances in the caller.